tests/Main.idr

```diff
@@ -1,12 +1,39 @@
 module Main
 
+import System
+import System.Directory
+import System.File
+
 import Test.Golden
 
 %default covering
 
 ------------------------------------------------------------------------
 -- Test cases
 
+isTest : (path : String) -> IO Bool
+isTest path = exists (path ++ "/run")
+
+filterTests : (testNames : List String) -> IO (List String)
+filterTests [] = pure []
+filterTests (n :: ns) =
+    do rem <- filterTests ns
+       case !(isTest n) of
+         True  => pure $ n :: rem
+         False => pure rem
+
+testsInDir : (dirName : String) -> (poolName : String) -> List Requirement -> Codegen -> IO TestPool
+testsInDir dirName poolName reqs cg = do
+  Right names <- listDir dirName
+    | Left e => do putStrLn ("failed to list " ++ dirName ++ ": " ++ show e)
+                   exitFailure
+  let testNames = [dirName ++ "/" ++ n | n <- names]
+  testNames <- filterTests testNames
+  when (length testNames == 0) $ do
+    putStrLn ("no tests found in " ++ dirName)
+    exitFailure
+  pure $ MkTestPool poolName reqs cg testNames
```

Comment:
This looks general enough to go into `Test.Golden` as part of the library.

Reply:
It is generic enough now (except that it calls `exitFailure` which might be not appropriate), but later I'd like to extend this code to discover options based on test content. E. g. read the list of codegens from the `.idr` file in the test.